### ISSUE TITLE
Update to JavaParser 3.18.0 from 3.17.0 causes NameExpr::resolve to hang

### ISSUE THREAD

**Issue by Col-E** [javaparser/javaparser#3071]
Possibly a duplicate of #3038 as I have only encountered this when explicitly resolving `NameExpr`

Attempting to resolve the `NameExpr` of `Foo` in the following code will take over an hour to resolve _(I stopped it after 75 minutes)_  if the set symbol solver has no way to access `Foo`

```java
public class Foo{
    public static void main(String[] args) {
        String s0   = "hello";
        String s1   = "hello";
        String s2   = "hello";
        String s3   = "hello";
        String s4   = "hello";
        String s5   = "hello";
        String s6   = "hello";
        String s7   = "hello";
        String s8   = "hello";
        String s9   = "hello";
        String s10  = "hello";
        String s11  = "hello";
        String s12  = "hello";
        String s13  = "hello";
        String s14  = "hello";
        String s15  = "hello";
        String s16  = "hello";
        String s17  = "hello";
        String s18  = "hello";
        String s19  = "hello";
        String s20  = "hello";
        String s21  = "hello";
        String s22  = "hello";
        String s23  = "hello";
        String s24  = "hello";
        String s25  = "hello";
        String s26  = "hello";
        String s27  = "hello";
        String s28  = "hello";
        String s29  = "hello";
        String s30  = "hello";
        String s31  = "hello";
        String s32  = "hello";
        String s33  = "hello";
        String s34  = "hello";
        String s35  = "hello";
        String s36  = "hello";
        String s37  = "hello";
        String s38  = "hello";
        String s39  = "hello";
        String s40  = "hello";
        String s41  = "hello";
        String s42  = "hello";
        String s43  = "hello";
        String s44  = "hello";
        String s45  = "hello";
        String s46  = "hello";
        String s47  = "hello";
        String s48  = "hello";
        String s49  = "hello";
        String s50  = "hello";
        String s51  = "hello";
        String s52  = "hello";
        String s53  = "hello";
        String s54  = "hello";
        String s55  = "hello";
        String s56  = "hello";
        String s57  = "hello";
        String s58  = "hello";
        String s59  = "hello";
        String s60  = "hello";
        String s61  = "hello";
        String s62  = "hello";
        String s63  = "hello";
        String s64  = "hello";
        String s65  = "hello";
        String s66  = "hello";
        String s67  = "hello";
        String s68  = "hello";
        String s69  = "hello";
        String s70  = "hello";
        String s71  = "hello";
        String s72  = "hello";
        String s73  = "hello";
        String s74  = "hello";
        String s75  = "hello";
        String s76  = "hello";
        String s77  = "hello";
        String s78  = "hello";
        String s79  = "hello";
        String s80  = "hello";
        String s81  = "hello";
        String s82  = "hello";
        String s83  = "hello";
        String s84  = "hello";
        String s85  = "hello";
        String s86  = "hello";
        String s87  = "hello";
        String s88  = "hello";
        String s89  = "hello";
        String s90  = "hello";
        String s91  = "hello";
        String s92  = "hello";
        String s93  = "hello";
        String s94  = "hello";
        String s95  = "hello";
        String s96  = "hello";
        String s97  = "hello";
        String s98  = "hello";
        String s99  = "hello";
        String s100 = "hello";
        new Thread(){
            @Override
            public void run() {
                Foo foo = Foo.getInstance(); // Resolve either Foo here
            }
        }.run();
    }
    static Foo getInstance() {
        return new Foo();
    }
}
```

This is due to what I believe is an oversight in `StatementContext#solveSymbol` where resolving the name checks each prior adjacent statement for a declaration of that name... but each time an adjacent prior statement is checked it tries to look in all of its prior adjacent statements. So if we have 5 adjacent statmeents, we dont visit just the 5. We visit the 5 then the 4 then the 3 then the 2 then the last 1.  This will cause an exponential amount of calls to `solveSymbol`.

**Comment by corradomio** [comment]
Exactly the same problem! With JP/JSS 3.19.1
I hack JSS to use a timeout to abort the resolution of the symbol.

A possible solution is to use "memoization" techniques:

https://en.wikipedia.org/wiki/Memoization

Because its implementation uses some GLOBAL/STATIC cache, it is useful to find a solution that permit the parallelism.

**Comment by Col-E** [comment]
> A possible solution is to use "memoization" techniques:

I am almost certain that this is just an oversight in the implemented algorithm. Check the linked PR. The included test with the fix runs perfectly fast and there are no other behavior regressions in the test suite.

**Comment by corradomio** [comment]
I don't understand: I am using JP/JSS 3.19.1. 
Is this version already patched?
I am analyzing Deeplearning4J and I have exactly the same problem.
However if you can specify the (more or less) line of code where is located the problem, I can add some extra log to identify the "new" case, not considered in the unit tests.

I found 

```public SymbolReference<? extends ResolvedValueDeclaration> solveSymbol(String name) {```

at line 191 of "StatementContext.java"

**Comment by Col-E** [comment]
> I don't understand: I am using JP/JSS 3.19.1.
> Is this version already patched?

JP 3.19 does not have this issue fixed. I'm saying that I've provided a fix in the linked PR #3072 and to me I think that the cause of this bug is simply an oversight in the resolving logic, not necessarily a slow algorithm in need of optimization via something like memoization.

**Comment by corradomio** [comment]
I understand ;-)
Little problem: you ""describe"" the solution, I have to ""implement"" it ;-)
If possible, obviously, can you post the changed part of code?
In this way I can implement the same fix ;-)

**Comment by Col-E** [comment]
Just click on the PR link and look at the _"[Files changed](https://github.com/javaparser/javaparser/pull/3072/files)"_ tab. It'll show you the git-diff. Its quite a small change.

**Comment by corradomio** [comment]
Thanks 1.000.000.000 ;-)

**Comment by Col-E** [comment]
To clarify I see that this issue seems to be classified as a _"performance problem"_ by the developers as indicated by this response: https://github.com/javaparser/javaparser/issues/2990#issuecomment-750402142

The linked PR in this issue shows that the same expected behavior of JavaParser can be achieved without this performance hit if we assume the underlying issue in 2290 is the same one described in this issue. The problem in this case/issue is that the same `NodeWithStmnt` is revisited in the resolve routine many times depending on the layout of the source code being analyzed. 

Its a niche case since most code does not have hundreds of lines in the same scoped level adjacent to one another as seen in the example code above, and requires resolving `NameExpr` instead of something more common like `MethodExpr`, 

I apologize if I'm being a bit pressing by cross checking and commenting on other issues with this one, but its a significant issue since it prevents me and other users from effectively updated to new versions of JavaParser. If anyone needs more information or an update to the code in the PR to get this closed out I would be happy to provide it.

**Comment by MysterAitch** [comment]
I think your description of the current behaviour being `O(n!)` is correct -- and is a mistake/needs to be corrected.

There are multiple "solve" methods within various "contexts", where the logic of where to search happens.

For example, here is the `StatementContext` where the "start at the current statement and work backwards" logic is:

https://www.github.com/javaparser/javaparser/tree/master/javaparser-symbol-solver-core%2Fsrc%2Fmain%2Fjava%2Fcom%2Fgithub%2Fjavaparser%2Fsymbolsolver%2Fjavaparsermodel%2Fcontexts%2FStatementContext.java

_Edit: just reread the posts above and noted that you're likely aware of this already - oops!_


To confirm, is this a regression where it works fine when using an older version but doesn't work fine now? 

I know I did some changes that was included within 3.19.0 so just wondering if that's where I should be looking.

_Edit 2: just reread the title... Seems I'm not having a great evening of reading comprehension 😅😅_


The linked PR seems reasonable though, so it's one way or another it'll be addressed within the next release.

**Comment by MysterAitch** [comment]
> To clarify I see that this issue seems to be classified as a _"performance problem"_ by the developers as indicated by this response: https://github.com/javaparser/javaparser/issues/2990#issuecomment-750402142


My understanding is that this comment was (rightly) stating that it is the correct behaviour happening very slowly -- as in, it'll give the right answer _eventually_ -- as a reason for removing the bug report label.

Happy to be corrected on that though if I've misread it!


_(...there is a whole other debate to be had about whether inefficient algorithms are bugs or not, but let's save that for another day :wink:)_
